File: cli/lib/method-manifest.js
/**
 * Loader for the shared method/provider registry SSOT (shared/method-registry.json).
 *
 * Single source of truth, shared with the Python arena harness
 * (arena/mt_eval_harness/method_manifest.py), for which translation METHODS
 * (MT engines) and LLM PROVIDERS exist and their declarative metadata.
 *
 * Mirrors lib/models.js's loader: prefer the package-bundled copy (cli/shared/,
 * shipped in the npm package), fall back to the monorepo-root SSOT for in-repo
 * dev. Returns null if neither is found, so the parity test skips rather than
 * failing in a context without the manifest.
 */

import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

let _cache; // undefined = not loaded, null = not found, object = loaded

/** @returns {object|null} parsed manifest, or null if not found */
export function loadMethodManifest() {
  if (_cache !== undefined) return _cache;
  const __dirname = path.dirname(fileURLToPath(import.meta.url));
  const candidates = [
    path.resolve(__dirname, '..', 'shared', 'method-registry.json'),
    path.resolve(__dirname, '..', '..', 'shared', 'method-registry.json'),
  ];
  for (const p of candidates) {
    try {
      _cache = JSON.parse(fs.readFileSync(p, 'utf-8'));
      return _cache;
    } catch { /* try next */ }
  }
  _cache = null;
  return null;
}

/**
 * Return manifest entries, optionally filtered to one kind.
 * @param {string|null} kind
 * @returns {Object<string, object>}
 */
export function manifestEntries(kind = null) {
  const m = loadMethodManifest();
  if (!m || !m.entries) return {};
  const out = {};
  for (const [name, entry] of Object.entries(m.entries)) {
    if (kind === null || entry.kind === kind) out[name] = entry;
  }
  return out;
}

/**
 * The CLI METHOD_REGISTRY key for a manifest entry: the cli_name override if
 * present (e.g. the canonical "openrouter" provider is "llm" in the CLI), else
 * the canonical name.
 */
export function cliNameFor(name, entry) {
  return (entry && entry.cli_name) || name;
}

File: arena/mt_eval_harness/method_manifest.py
"""
Loader for the shared method/provider registry SSOT (shared/method-registry.json).

This is the single source of truth, shared with the JS CLI, for which
translation METHODS (MT engines) and LLM PROVIDERS exist and their declarative
metadata (env vars, method_class, paradigm, license, cost, locale maps).

Mirrors ``config._load_model_aliases``: walk up from this package to find the
monorepo ``shared/`` directory. Unlike model-aliases there is NO embedded
standalone fallback — the manifest is purely a dev/CI SSOT-drift guard
(consumed by the parity tests), and a standalone pip install simply has no
manifest to check against (callers treat a missing manifest as "skip the
parity check", never as an error). The runtime registries
(methods/registry.py, providers/registry.py) do NOT hard-depend on this file,
so standalone installs keep working.
"""

from __future__ import annotations

import json
from pathlib import Path

_PACKAGE_DIR = Path(__file__).resolve().parent


def manifest_path() -> Path | None:
    """Return the path to shared/method-registry.json, or None if not found.

    Walks up from this package looking for the monorepo ``shared/`` dir.
    Returns None in a standalone install where ``shared/`` isn't present.
    """
    check = _PACKAGE_DIR
    for _ in range(6):
        candidate = check / "shared" / "method-registry.json"
        if candidate.exists():
            return candidate
        check = check.parent
    return None


def load_method_manifest() -> dict | None:
    """Load and return the parsed manifest dict, or None if not found.

    Raises ValueError if the file exists but is malformed (fail loud — a
    corrupt SSOT must never silently pass the parity guard).
    """
    path = manifest_path()
    if path is None:
        return None
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except json.JSONDecodeError as exc:
        raise ValueError(f"Malformed method-registry SSOT at {path}: {exc}") from exc


def manifest_entries(kind: str | None = None) -> dict[str, dict]:
    """Return manifest entries, optionally filtered to one ``kind``.

    Returns {} if the manifest isn't found (standalone install).
    """
    manifest = load_method_manifest()
    if not manifest:
        return {}
    entries = manifest.get("entries", {})
    if kind is None:
        return dict(entries)
    return {name: e for name, e in entries.items() if e.get("kind") == kind}


def manifest_names(kind: str | None = None) -> set[str]:
    """Return the set of entry names, optionally filtered to one ``kind``."""
    return set(manifest_entries(kind).keys())
